module/Database/src/Mapper/ProspectiveMember.php

```diff
@@ -96,6 +124,41 @@ public function find(int $lidnr): ?ProspectiveMemberModel
         return $qb->getQuery()->getOneOrNullResult();
     }
 
+    /**
+     * Find all prospective members whose last Checkout Session has fully expired ((1 + 30) + 1 day ago) or failed 31
+     * days ago.
+     *
+     * @return ProspectiveMemberModel[]
+     */
+    public function findWithFullyExpiredOrFailedCheckout(): array
+    {
+        $qb = $this->getRepository()->createQueryBuilder('m');
+        $qb->leftJoin(CheckoutSessionModel::class, 'cs', Join::WITH, 'cs.prospectiveMember = m.lidnr');
+
+        $qbc = $this->em->createQueryBuilder();
+        $qbc->select('MAX(css.id)')
```

Comment:
Here in the code we assume that the latest checkoutsession is always correct, which has the underlying assumption that we do not create a new one before the previous one is no longer usable. I think this works properly, but I am not completely sure.

Reply:
This is only problematic if Stripe delays a `checkout.session.expired` and does not destroy it if in the mean time `checkout.session.completed` was emitted. I do not see a way to solve this, things like `updatedAt` or `created(At)` will still be wrong on that case.

Technically, you could prioritise a `paid` over `pending` over `failed`/`expired` over `created` but then we have to redo the ordering in `CheckoutSessionStates`. But this is problematic in itself; what if a payment fails, the user starts a new session and subsequently lets it expire. Then we would also have to sort on the `created` in that case.

This complicates things a lot and will make our lives miserable. I _think_ that we can safely assume that Stripe does its job (nearly) perfect.

Reply:
Let's do that indeed